features/flows/elastic/src/main/java/org/opennms/netmgt/flows/elastic/thresholding/FlowThresholding.java

```diff
@@ -133,78 +123,144 @@ public void run() {
 
     public void threshold(final List<FlowDocument> documents,
                           final FlowSource source) throws ExecutionException, ThresholdInitializationException {
 
+        final var now = Instant.now();
+
         for (final var document : documents) {
             if (document.getNodeExporter() != null && !Strings.isNullOrEmpty(document.getApplication())) {
                 final var exporterKey = new ExporterKey(document.getNodeExporter().getInterfaceId());
 
-                final var applicationKey = new ApplicationKey(exporterKey,
-                        document.getDirection() == Direction.INGRESS
-                                ? document.getInputSnmp()
-                                : document.getOutputSnmp(),
-                        document.getApplication());
+                final var session = this.sessions.computeIfAbsent(exporterKey, key -> {
+                    LOG.debug("Accepting session for {}", exporterKey);
+
+                    final OnmsIpInterface iface = this.ipInterfaceDao.get(exporterKey.interfaceId);
+
+                    final CollectionAgent collectionAgent = FlowThresholding.this.collectionAgentFactory.createCollectionAgent(iface);
+
+                    final ThresholdingSession thresholdingSession;
+                    try {
+                        thresholdingSession = FlowThresholding.this.thresholdingService.createSession(iface.getNodeId(),
+                                                                                                      collectionAgent.getHostAddress(),
+                                                                                                      SERVICE_NAME,
+                                                                                                      FLOW_APP_RRD_REPO,
+                                                                                                      new ServiceParameters(Collections.emptyMap()));
+                    } catch (ThresholdInitializationException e) {
+                        throw new RuntimeException("Error initializing thresholding session", e);
+                    }
+
+                    return new Session(thresholdingSession,
+                                       collectionAgent);
+                });
 
-                this.applicationAccumulator.computeIfAbsent(applicationKey, k -> new AtomicLong(0)).addAndGet(document.getBytes());
+                session.process(now, document);
             }
         }
     }
 
     private static class Session {
+        public final Map<ApplicationKey, AtomicLong> applications;
+
         public final ThresholdingSession thresholdingSession;
         public final CollectionAgent collectionAgent;
 
+        // The last time this session sees any incoming flow.
+        // This is not synchronized as reference updates are always atomic.
+        // See https://docs.oracle.com/javase/specs/jls/se7/html/jls-17.html#jls-17.7
+        private Instant lastUpdate = null;
```

Comment:
Should be volatile, otherwise timer task may not see last value.

Reply:
Same as above. This is only accessed once in the method.

Reply:
As said, `volatile` has nothing to do with methods.

Reply:
altered.